src/streaming/zarr.common.cpp

```diff
@@ -0,0 +1,297 @@
+#include "macros.hh"
+#include "zarr.common.hh"
+
+#include <stdexcept>
+
+size_t
+zarr::bytes_of_type(ZarrDataType data_type)
+{
+    switch (data_type) {
+        case ZarrDataType_int8:
+        case ZarrDataType_uint8:
+            return 1;
+        case ZarrDataType_int16:
+        case ZarrDataType_uint16:
+            return 2;
+        case ZarrDataType_int32:
+        case ZarrDataType_uint32:
+        case ZarrDataType_float32:
+            return 4;
+        case ZarrDataType_int64:
+        case ZarrDataType_uint64:
+        case ZarrDataType_float64:
+            return 8;
+        default:
+            throw std::runtime_error("Invalid data type: " +
+                                     std::to_string(data_type));
+    }
+}
+
+size_t
+zarr::bytes_of_frame(const std::vector<Dimension>& dims, ZarrDataType type)
+{
+    return bytes_of_type(type) * dims.back().array_size_px *
+           dims[dims.size() - 2].array_size_px;
+}
+
+size_t
+zarr::chunks_along_dimension(const Dimension& dimension)
+{
+    EXPECT(dimension.chunk_size_px > 0, "Invalid chunk size.");
+
+    return (dimension.array_size_px + dimension.chunk_size_px - 1) /
+           dimension.chunk_size_px;
+}
+
+size_t
+zarr::shards_along_dimension(const Dimension& dimension)
+{
+    const size_t shard_size = dimension.shard_size_chunks;
+    if (shard_size == 0) {
+        return 0;
+    }
+
+    const size_t n_chunks = chunks_along_dimension(dimension);
+    return (n_chunks + shard_size - 1) / shard_size;
+}
+
+size_t
+zarr::chunk_lattice_index(size_t frame_id,
+                          size_t dimension_idx,
+                          const std::vector<Dimension>& dims)
+{
+    // the last two dimensions are special cases
+    EXPECT(dimension_idx < dims.size() - 2,
+           "Invalid dimension index: %zu",
+           dimension_idx);
+
+    // the first dimension is a special case
```

Comment:
To improve safety and maintainability when dealing with special cases in our dimension handling, we could consider creating a dedicated type that provides a robust interface for accessing dimension data. This approach would allow us to encapsulate the logic associated with special cases, giving them meaningful names that convey their purpose. By doing so, we can provide bounded access to the non-special case dimension data, reducing the risk of errors and enhancing code readability. Such an interface would promote clearer usage patterns and facilitate easier updates in the future, as any changes to special case handling would be centralized within this type, rather than scattered throughout the codebase.

I think this is important since this knowledge is spread across multiple functions.

Reply:
Yeah, I think this is a fine idea.